src/mrinufft/operators/off_resonance.py

```diff
@@ -0,0 +1,428 @@
+        assert CUPY_AVAILABLE, "GPU computation requires Cupy!"
+        xp = cp
+        b0_map = _to_cupy(b0_map)
+        readout_time = _to_cupy(readout_time)
+        mask = _to_cupy(mask)
+        r2star_map = _to_cupy(r2star_map)
+    else:
+        xp = np
+        b0_map = _to_numpy(b0_map)
+        readout_time = _to_numpy(readout_time)
+        mask = _to_numpy(mask)
+        r2star_map = _to_numpy(r2star_map)
+
+    readout_time = xp.asarray(readout_time, dtype=xp.float32).ravel()
+    if mask is None:
+        mask = xp.ones_like(b0_map, dtype=bool)
+    else:
+        mask = xp.asarray(mask, dtype=bool)
+
+    # Hz to radians / s
+    field_map = _get_complex_fieldmap(b0_map, r2star_map)
+
+    # enforce precision
+    field_map = xp.asarray(field_map, dtype=xp.complex64)
+
+    # create histograms
+    z = field_map[mask].ravel()
+
+    if r2star_map is not None:
+        z = xp.stack((z.imag, z.real), axis=1)
+        hk, ze = xp.histogramdd(z, bins=n_bins)
+        ze = list(ze)
+
+        # get bin centers
+        zc = [e[1:] - (e[1] - e[0]) / 2 for e in ze]
+
+        # complexify
+        zk = _outer_sum(1j * zc[0], zc[1])  # [K1 K2]
+        zk = zk.T
+        hk = hk.T
+    else:
+        hk, ze = xp.histogram(z.imag, bins=n_bins[0])
+
+        # get bin centers
+        zc = ze[1:] - (ze[1] - ze[0]) / 2
+
+        # complexify
+        zk = 1j * zc  # [K 1]
+
+    # flatten histogram values and centers
+    hk = hk.ravel()
+    zk = zk.ravel()
+
+    # generate time for each segment
+    tl = xp.linspace(
+        readout_time.min(), readout_time.max(), n_time_segments, dtype=xp.float32
+    )  # time seg centers in [s]
+
+    # prepare for basis calculation
+    ch = xp.exp(-tl[:, None, ...] @ zk[None, ...])
+    w = xp.diag(hk**0.5)
+    p = xp.linalg.pinv(w @ ch.T) @ w
+
+    # actual temporal basis calculation
+    B = p @ xp.exp(-zk[:, None, ...] * readout_time[None, ...])
+    B = B.astype(xp.complex64)
+
+    # back to torch if required
+    if is_torch:
+        B = _to_torch(B)
+        tl = _to_torch(tl)
+
+    return B, tl
+
+
+def _outer_sum(xx, yy):
+    xx = xx[:, None, ...]  # add a singleton dimension at axis 1
+    yy = yy[None, ...]  # add a singleton dimension at axis 0
+    ss = xx + yy  # compute the outer sum
+    return ss
+
+
+# TODO: /* refactor with_* decorators
+def _to_numpy(input):
+    if input is None:
+        return input
+    xp = get_array_module(input)
+
+    if xp.__name__ == "torch":
+        return input.numpy(force=True)
+    elif xp.__name__ == "cupy":
+        return input.get()
+    else:
+        return input
+
+
+def _to_cupy(input):
+    if input is None:
+        return input
+    return cp.asarray(input)
+
+
+def _to_torch(input):
+    xp = get_array_module(input)
+
+    if xp.__name__ == "numpy":
+        return torch.from_numpy(input)
+    elif xp.__name__ == "cupy":
+        return torch.from_dlpack(input)
+    else:
+        return input
+
+
+# */
```

Comment:
What is that ?

Reply:
I just wanted to mark the block I will remove after the with_* decorators (e.g., with_torch) will be generalized (I have done some work on this and will open another PR after this one is ok)